### PR TITLE
mobile: iOS bluetooth low energy

### PR DESCRIPTION
**Pull request by knguyenuit**
#### What's this PR do?

- [x] Basic knowledge about the bluetooth low energy in iOS
- [x] Demo connect the iOS device with the Bluetooth device and transfer data

### CODE REVIEW

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    <key>NSBluetoothAlwaysUsageDescription</key>
+    <string>We use Bluetooth to show basic communication between Central and Peripheral</string>
+    <key>NSBluetoothPeripheralUsageDescription</key>
+    <string>We use Bluetooth to show basic communication between Central and Peripheral</string>
+    <key>UIApplicationSupportsIndirectInputEvents</key>
+```
+
+* Init BLE
+
+```
```
Comment (monotykamary) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:97]: Small nit here for code visibility:
\```swift

**Review by monotykamary — CHANGES_REQUESTED** [review]
A great entry, with some small nits such as the duplicate markdown entry, lack of frontmatter, and lack of code highlights. Not too sure how to handle the demo zip, but I think we can organize it on our side.

File: .gitignore [hunk 2]
```diff
@@ -20,4 +20,6 @@
 # debug
 npm-debug.log*
 yarn-debug.log*
-yarn-error.log*
\ No newline at end of file
+yarn-error.log*
+knguyenuit_rsa
+knguyenuit_rsa.pub
```
Comment (monotykamary) [inline .gitignore:25]: Why is this included in the gitignore?

**Review by monotykamary — CHANGES_REQUESTED** [review]

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    case connectedDiscovering
+    case connected
+    }
+```
+
+When scan and connect the central to peripheral, implement count down 30s timeout.
+
+* Count down time
+
+```
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:121]: Is this swift language? If so, can you specify `lang` of the block for the sake of sytax highlight?

```suggestion
```swift
```

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+        }
+        guard let data = textJson.data(using: .utf8),
+              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
+              let result = json["result"] as? String
+        else {
+            showAlert(title: "Data Send", message: "FAIL")
+            return
+        }
+        showAlert(title: "Data Send", message: result == "ng" ? "FAIL" : "SUCCESS")
+    }
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:229]: I think we can remove the leading space of the whole block
```suggestion
private func handleDataNotify(_ data: Data) {
    var textJson = String(data: data, encoding: .utf8) ?? ""
    if let indexRemove = textJson.firstIndex(where: { $0 == "<" }) {
        textJson.remove(at: indexRemove)
    }
    if let indexRemove = textJson.firstIndex(where: { $0 == ">" }) {
        textJson.remove(at: indexRemove)
    }
    guard let data = textJson.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let result = json["result"] as? String
    else {
        showAlert(title: "Data Send", message: "FAIL")
        return
    }
    showAlert(title: "Data Send", message: result == "ng" ? "FAIL" : "SUCCESS")
}
```

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+        connectedPeripheral?.writeValue(data, for: characteristic, type: .withResponse)
+    }
+```
+
+Convert and handle data send to the Peripheral and receive from Peripheral
+
+* Data send
+
+```
+    let text = textFieldDataForWrite.text ?? ""
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:204]: Same here, can you remove the leading space?

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+            handleDataNotify(data)
+        }
+        appendLog("didUpdateValue '\(stringValue)'")
+    }
+```
+
+* Function write and read Characteristic
+
+```
+    func bleReadCharacteristic(uuid: CBUUID) {
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:182]: Same here, can you remove the leading space and specify the language for the block?

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+            }
+
+        }
+    }
+```
+
+* Handle data recevice from bluetooth device (peripheral)
+
+```
+    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:164]: Same here, can you remove the leading space and specify the language for the block?

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+```
+    private func initBLE() {
+        // using DispatchQueue.main means we can update UI directly from delegate methods
+        bleCentral = CBCentralManager(delegate: self, queue: DispatchQueue.main)
+    }
+```
+
+* BLELifecycleState
+
+```
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:106]: Same here, can you remove the leading space and specify the language for the block?

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+        let textSuccess = "<{\"action\":\"bleconfig\",\"result\":\"ok\"}>"
+        let textFail = "<{\"action\":\"bleconfig\",\"result\":\"ng\"}>"
+        bleSendIndication(isResponseSuccess ? textSuccess : textFail)
+    }
+```
+
+* Function didReceiveWrite will receive data from BLE Central
+
+```
+    func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:330]: Same here, can you remove the leading space and specify the language for the block?

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+            }
+        }
+        appendLog(log)
+    }
+```
+
+* When receive data from BLE Central, BLE Prepheral will notify success or fail to BLE Central
+
+```
+    private func bleSendIndication(_ valueString: String) {
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:355]: Same here, can you remove the leading space and specify the language for the block?

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+    }
+```
+
+## Recommended resources
+
+iOS:
+
+* [Transferring Data Between Bluetooth Low Energy Devices](https://developer.apple.com/documentation/corebluetooth/transferring_data_between_bluetooth_low_energy_devices) by Apple (with sample code, Central and Peripheral)
+* [The Ultimate Guide to Apple’s Core Bluetooth](https://punchthrough.com/core-bluetooth-basics/) by PunchThrough
+* [Core Bluetooth Tutorial](https://www.raywenderlich.com/231-core-bluetooth-tutorial-for-ios-heart-rate-monitor) by Raywenderlich
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:373]:
```suggestion
- [Transferring Data Between Bluetooth Low Energy Devices](https://developer.apple.com/documentation/corebluetooth/transferring_data_between_bluetooth_low_energy_devices) by Apple (with sample code, Central and Peripheral)
- [The Ultimate Guide to Apple’s Core Bluetooth](https://punchthrough.com/core-bluetooth-basics/) by PunchThrough
- [Core Bluetooth Tutorial](https://www.raywenderlich.com/231-core-bluetooth-tutorial-for-ios-heart-rate-monitor) by Raywenderlich
```

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+                                                      properties: .indicate,
+                                                      value: nil,
+                                                      permissions: .readable)
+        self.charForIndicate = charForIndicate
+
+        // create service
+        let service = CBMutableService(type: uuidService, primary: true)
+        service.characteristics = [charForRead, charForWrite, charForIndicate]
+        return service
+    }
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:292]: Can we format this block of code?

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+            return
+        }
+        sendNotify(isResponseSuccess: pass == "1")
+    }
+    
+    private func sendNotify(isResponseSuccess: Bool) {
+        let textSuccess = "<{\"action\":\"bleconfig\",\"result\":\"ok\"}>"
+        let textFail = "<{\"action\":\"bleconfig\",\"result\":\"ng\"}>"
+        bleSendIndication(isResponseSuccess ? textSuccess : textFail)
+    }
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:324]: Can we format this block of code?

File: Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+            showAlert(title: "Data Send", message: "FAIL")
+            return
+        }
+        showAlert(title: "Data Send", message: result == "ng" ? "FAIL" : "SUCCESS")
+    }
+```
+
+* Define format data send to the Peripheral
+
+```
```
Comment (zlatanpham) [inline Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md:234]:
```suggestion
```swift
```

**Review by zlatanpham — COMMENTED** [review]
Left some comments about the formatting.

**Comment by zlatanpham** [comment]
@viettrungphan can you take a look too?

**Comment by viettrungphan** [comment]
There are some things should not in the sample code related to security, I will discuss with him separately and guide him how to fix it. Do not merge. Thanks @zlatanpham @monotykamary.

**Pull request closed without merging**